### PR TITLE
Implement indexing operations in PyTorch

### PR DESCRIPTION
**Pull request by HarshvirSandhu**
Implement indexing ops in pytorch

- [x] Subtensor
- [x] IncSubtensor
- [x] AdvancedSubtensor[1]
- [x] AdvancedIncSubtensor[1]

## Related Issue

- [x] Related to #821 

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works

## Type of change

- [x] New feature / enhancement

### CODE REVIEW

File: tests/link/pytorch/test_subtensor.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    # assert isinstance(out_pt.owner.op, pt_subtensor.AdvancedSubtensor)
+    # out_fg = FunctionGraph([x_pt], [out_pt])
+    # compare_pytorch_and_py(out_fg, [x_np])
+
+    # Flipping
+    # out_pt = x_pt[::-1]
+    # out_fg = FunctionGraph([x_pt], [out_pt])
+    # compare_pytorch_and_py(out_fg, [x_np])
+
+    # Boolean indexing should work if indexes are constant
```
Comment (ricardoV94) [inline tests/link/pytorch/test_subtensor.py:62]:
```suggestion
```

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            if isinstance(i, slice):
+                if i.start:
+                    new_start = i.start.item()
+                if i.step:
+                    new_step = i.step.item()
+                if i.stop:
+                    new_stop = i.stop.item()
+                new_indices.append(slice(new_start, new_stop, new_step))
+            else:
+                new_indices.append(i.tolist())
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:31]: Why is `.tolist()` needed, and what happens if there's a 2d index?

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        for i in indices:
+            new_start, new_step, new_stop = None, None, None
+            if isinstance(i, slice):
+                if i.start:
+                    new_start = i.start.item()
+                if i.step:
+                    new_step = i.step.item()
+                if i.stop:
+                    new_stop = i.stop.item()
+                new_indices.append(slice(new_start, new_stop, new_step))
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:29]: Why is this needed? I think the slices should be scalars already?

Comment (HarshvirSandhu) [reply pytensor/link/pytorch/dispatch/subtensor.py:29]: `start`, `step`, `stop` are pytorch tensors (example: `slice(tensor(1), tensor(2), None)`) and have to be converted to scalars

File: pytensor/link/pytorch/dispatch/subtensor.py (hunk 2 above)
Comment (HarshvirSandhu) [reply pytensor/link/pytorch/dispatch/subtensor.py:31]: Have not tried with a 2d index yet, but `.tolist` is needed for such cases:
`pt_subtensor.advanced_subtensor1(x_pt, [1, 2])`

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            else:
+                new_indices.append(i.tolist())
+
+        if len(indices) == 1:
+            indices = indices[0]
+
+        return x[*new_indices]
```
[the remark this answers is not in the document]
Comment (HarshvirSandhu) [reply pytensor/link/pytorch/dispatch/subtensor.py:36]: yes, this works

File: tests/link/pytorch/test_subtensor.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    # out_pt = x_pt[[1, 2], :, [3, 4]]
+    # assert isinstance(out_pt.owner.op, pt_subtensor.AdvancedSubtensor)
+    # out_fg = FunctionGraph([x_pt], [out_pt])
+    # compare_pytorch_and_py(out_fg, [x_np])
+
+    # Flipping
+    # out_pt = x_pt[::-1]
+    # out_fg = FunctionGraph([x_pt], [out_pt])
+    # compare_pytorch_and_py(out_fg, [x_np])
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply tests/link/pytorch/test_subtensor.py:60]: Then we should at least raise `NotImplementedError` when at runtime the steps are negative

File: tests/link/pytorch/test_subtensor.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
+import numpy as np
+
+from pytensor.graph.fg import FunctionGraph
+from pytensor.tensor import subtensor as pt_subtensor
+from pytensor.tensor import tensor
+from tests.link.pytorch.test_basic import compare_pytorch_and_py
+
+
+def test_pytorch_Subtensor():
```
[the remark this answers is not in the document]
Comment (HarshvirSandhu) [reply tests/link/pytorch/test_subtensor.py:10]: This fails with this error:
```
>       return torch.as_tensor(data, dtype=dtype)
E       RuntimeError: Could not infer dtype of NoneType
```

`pytorch_funcify` gets non array data, the same happens when  we use `x_pt[[1, 2], :, [3, 4]]` in this case `pytorch_funcify` gets a slice object

[the remark this answers is not in the document]
Comment (ricardoV94) [reply tests/link/pytorch/test_subtensor.py:10]: We need to tipyfy Constant None Variables and  Slices. Our default assumes everything is an array. Actually the default should raise NotImplementedError like we do for funcify

[the remark this answers is not in the document]
Comment (HarshvirSandhu) [reply tests/link/pytorch/test_subtensor.py:10]: Could we do something like this for typify?
```
def pytorch_typify(data, dtype=None, **kwargs):
        if data is None or isinstance(data, slice):
                return data
        return torch.as_tensor(data, dtype=dtype)
```

[the remark this answers is not in the document]
Comment (ricardoV94) [reply tests/link/pytorch/test_subtensor.py:11]: None is being typified in the PR that adds shape Ops to PyTorch, check how it looks there. Should be the same here as that will likely be merged before this one

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+    def subtensor(x, *ilists):
+        indices = indices_from_subtensor(ilists, idx_list)
+        new_indices = []
+
+        for i in indices:
+            new_start, new_step, new_stop = None, None, None
+            if isinstance(i, slice):
+                if i.start:
+                    new_start = i.start.item()
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:26]: I think this is not needed if we don't typify integers and tensor by default. I think they are originally integers?

File: pytensor/link/pytorch/dispatch/basic.py [hunk 8]
```diff
@@ -13,7 +13,10 @@
 @singledispatch
 def pytorch_typify(data, dtype=None, **kwargs):
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/basic.py:14]: We should make the default `pytorch_typify` raise NotImplementedError, and then have the one that was covering the `torch.as_tensor` cases dispatched on `np.ndarray` and torch tensors

File: pytensor/link/pytorch/dispatch/basic.py [hunk 9]
```diff
@@ -13,7 +13,12 @@
...
     r"""Convert instances of PyTensor `Type`\s to PyTorch types."""
-    return torch.as_tensor(data, dtype=dtype)
+    try:
+        return torch.as_tensor(data, dtype=dtype)
+    except RuntimeError:
+        raise NotImplementedError(
+            f"pytorch_typify got type {type(data)}, it should be an array"
+        )
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/basic.py:21]: This is not what I meant. The default should just raise NotImplementedError.

Then the old default of `torch.as_tensor` should be a function we dispatch for tensor/ndarray types, like we dispatch the None function on NoneType

File: pytensor/link/pytorch/dispatch/subtensor.py (hunk 2 above)
Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/subtensor.py:30]: Torch doesn't accept tensors as indices? Does it require converting them to a list? Otherwise we shouldn't have to to be doing this conversion

Comment (HarshvirSandhu) [reply pytensor/link/pytorch/dispatch/subtensor.py:30]: There is a shape mismatch when passing tensors as indices

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+    if getattr(op, "set_instead_of_inc", False):
+
+        def torch_fn(x, indices, y):
+            return x.at[indices].set(y)
+
+    else:
+
+        def torch_fn(x, indices, y):
+            return x.at[indices].add(y)
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:53]: This looks like jax syntax, not pytorch? Are you sure about it?

File: pytensor/link/pytorch/dispatch/basic.py [hunk 11]
```diff
@@ -13,7 +13,12 @@
 @singledispatch
 def pytorch_typify(data, dtype=None, **kwargs):
     r"""Convert instances of PyTensor `Type`\s to PyTorch types."""
-    return torch.as_tensor(data, dtype=dtype)
+    if isinstance(data, NoneType):
+        raise NotImplementedError(f"pytorch_typify is not implemented for {type(data)}")
+    if isinstance(data, slice):
+        raise NotImplementedError(f"pytorch_typify is not implemented for {type(data)}")
+    else:
+        return torch.as_tensor(data, dtype=dtype)
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/basic.py:21]: You're missing the idea. By default nothing is implemented:

```python
@singledispatch
def pytorch_typify(data, **kwargs):
  raise NotImplementedError(f"pytorch_typify is not implemente for {type(data)}")
```

Then you can have a specific dispatch for the cases that were previously covered by the default case

```python
@pytorch_typify.register(np.ndarray)
@pytroch.typify.register(torch.Tensor)
def pytorch_tipyfy_tensor(data, dtype=None, **kwargs):
  return torch.as_tensor(data, dtype=dtype)
```

Does that make sense?

File: pytensor/link/pytorch/dispatch/subtensor.py (hunk 2 above)
Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/subtensor.py:30]: Can you provide more details? When I run the following code it works fine:

```python
import torch
x = torch.arange(5*3).reshape((5, 3))
x[torch.tensor([0, 2, 4])]
```
I did not need to convert the tensor indices with .tolist(), which should be the same case here?

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+                if i.step and i.step < 0:
+                    raise NotImplementedError(
+                        "Negative step sizes are not supported in Pytorch"
+                    )
+                new_indices.append(i)
+            else:
+                new_indices.append(i.tolist())
+
+        if len(indices) == 1:
+            indices = indices[0]
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:33]: This seems wrong

File: tests/link/pytorch/test_subtensor.py [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+    out_fg = FunctionGraph([x_pt], [out_pt])
+    compare_pytorch_and_py(out_fg, [x_np])
+
+    # Advanced and basic indexing
+    out_pt = x_pt[[1, 2], :]
+    assert isinstance(out_pt.owner.op, pt_subtensor.AdvancedSubtensor)
+    out_fg = FunctionGraph([x_pt], [out_pt])
+    compare_pytorch_and_py(out_fg, [x_np])
+
+    with pytest.raises(NotImplementedError):
```
Comment (ricardoV94) [inline tests/link/pytorch/test_subtensor.py:54]: More details on why this is not Implemented?

File: tests/link/pytorch/test_subtensor.py [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+        out_pt = x_pt[[1, 2], :, [3, 4]]
+        assert isinstance(out_pt.owner.op, pt_subtensor.AdvancedSubtensor)
+        out_fg = FunctionGraph([x_pt], [out_pt])
+        compare_pytorch_and_py(out_fg, [x_np])
+
+    # Flipping
```
Comment (ricardoV94) [inline tests/link/pytorch/test_subtensor.py:60]: Negative steps are more than just flipping, I suggest removing the label or renaming it to `# negative step`
```suggestion
```

File: tests/link/pytorch/test_subtensor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ricardoV94) [reply tests/link/pytorch/test_subtensor.py:11]: Did you figure out `None`? Should already be merged I think

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

File: pytensor/link/pytorch/dispatch/subtensor.py (hunk 2 above)
Comment (HarshvirSandhu) [reply pytensor/link/pytorch/dispatch/subtensor.py:30]: In this case the indices are torch tensors, we have a list of tensors instead of a one dimensional tensor

Reproducible code:
```
import torch
x = torch.arange(5*3*5*4).reshape((5, 3, 5, 4))
a = x[torch.tensor([0, 2, 4])]

indices = (torch.tensor(0), torch.tensor(2), torch.tensor(4))

b = x[indices]

print(a.shape) # torch.Size([3, 3, 5, 4])
print(b.shape) # torch.Size([4])
```

Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/subtensor.py:30]: Sure but that also doesn't require any call to .tolist()

Comment (HarshvirSandhu) [reply pytensor/link/pytorch/dispatch/subtensor.py:30]: We could use `.item` instead of `.tolist` but that would not work cases like: `pt.subtensor.advanced_subtensor1(x, [1, 2])`

Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/subtensor.py:30]: I think we need to take a step back here. This should be simpler. 

First let's separate Subtensor from AdvancedSubtensor.

Subtensor only allows basic indexing and no array indices. This shouldn't need any type conversion.

Then AdvancedSubtensor allows both, but distinctively you don't need to create slice objects as those are inputs to the Op, unlike the basic Subtensor where you have to create them as the inputs are just start, stop and step of the slices.

File: pytensor/link/pytorch/dispatch/basic.py [hunk 15]
```diff
@@ -1,24 +1,40 @@
...
+def pytorch_typify_tensor(data, dtype=None, **kwargs):
     return torch.as_tensor(data, dtype=dtype)
 
 
+@pytorch_typify.register(slice)
 @pytorch_typify.register(NoneType)
-def pytorch_typify_None(data, **kwargs):
-    return None
+@pytorch_typify.register(np.number)
+def pytorch_typify_scalar(data, **kwargs):
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/basic.py:36]:
```suggestion
def pytorch_typify_no_conversion_needed(data, **kwargs):
```

File: pytensor/link/pytorch/dispatch/basic.py [hunk 16]
```diff
@@ -132,3 +148,11 @@ def makevector(*x):
         return torch.tensor(x, dtype=torch_dtype)
 
     return makevector
+
+
+@pytorch_funcify.register(TensorFromScalar)
+def pytorch_funcify_TensorFromScalar(op, **kwargs):
+    def tensorfromscalar(x):
+        return x
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/basic.py:156]: The Op should create a tensor
```suggestion
        return torch.as_tensor(x)
```

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    idx_list = getattr(op, "idx_list", None)
+
+    def subtensor(x, *ilists):
+        indices = indices_from_subtensor(ilists, idx_list)
+        for i in indices:
+            if isinstance(i, slice):
+                if i.step and i.step < 0:
+                    raise NotImplementedError(
+                        "Negative step sizes are not supported in Pytorch"
+                    )
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:24]: Make this checker a helper since it's needed for all Subtensor (basic, advanced, inc) Ops

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+@pytorch_funcify.register(MakeSlice)
+def pytorch_funcify_makeslice(op, **kwargs):
+    def makeslice(*x):
+        return slice(x)
+
+    return makeslice
+
+
+@pytorch_funcify.register(IncSubtensor)
+@pytorch_funcify.register(AdvancedIncSubtensor1)
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:55]: Split IncSubtensor (basic indexing) and AdvancedIncSubtensor (advanced indexing) like you did for Subtensor and AdvancedSubtensor. Also why is AdvavcedSubtensor (without 1) not covered?

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+def pytorch_funcify_IncSubtensor(op, node, **kwargs):
+    idx_list = getattr(op, "idx_list", None)
+
+    if getattr(op, "set_instead_of_inc", False):
+
+        def torch_fn(x, indices, y):
+            x[indices] = y
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:62]: Should check for negative steps in slices, like we do for indexing

File: tests/link/pytorch/test_subtensor.py [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+    # Negative step
+    with pytest.raises(
+        NotImplementedError, match="Negative step sizes are not supported in Pytorch"
+    ):
+        out_pt = x_pt[::-1]
+        out_fg = FunctionGraph([x_pt], [out_pt])
+        compare_pytorch_and_py(out_fg, [x_np])
+
+
+def test_pytorch_AdvSubtensor():
```
Comment (ricardoV94) [inline tests/link/pytorch/test_subtensor.py:56]: Forgot to test error with negative steps

File: tests/link/pytorch/test_subtensor.py [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+    assert isinstance(out_pt.owner.op, pt_subtensor.IncSubtensor)
+    out_fg = FunctionGraph([], [out_pt])
+    compare_pytorch_and_py(out_fg, [])
+
+    out_pt = pt_subtensor.set_subtensor(x_pt[0, 1:3, 0], st_pt)
+    assert isinstance(out_pt.owner.op, pt_subtensor.IncSubtensor)
+    out_fg = FunctionGraph([], [out_pt])
+    compare_pytorch_and_py(out_fg, [])
+
+    # "Set" advanced indices
```
Comment (ricardoV94) [inline tests/link/pytorch/test_subtensor.py:128]: Split AdvancedSetSubtensor cases into their own test

File: tests/link/pytorch/test_subtensor.py [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+
+    x_np = rng.uniform(-1, 1, size=(3, 4, 5)).astype(config.floatX)
+    x_pt = pt.constant(np.arange(3 * 4 * 5).reshape((3, 4, 5)).astype(config.floatX))
+
+    # "Set" basic indices
+    st_pt = pt.as_tensor_variable(np.array(-10.0, dtype=config.floatX))
+    out_pt = pt_subtensor.set_subtensor(x_pt[1, 2, 3], st_pt)
+    assert isinstance(out_pt.owner.op, pt_subtensor.IncSubtensor)
+    out_fg = FunctionGraph([], [out_pt])
+    compare_pytorch_and_py(out_fg, [])
```
Comment (ricardoV94) [inline tests/link/pytorch/test_subtensor.py:115]: Don't test cases that are constant (don't need any `numpy` arguments in `compare_pytorch_and_py`), because we may not end up testing the implementation in pytorch at all

**Review by ricardoV94 — CHANGES_REQUESTED** [review]
This is looking pretty good! Some change requests and suggestions to make the tests more robust.

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+
+@pytorch_funcify.register(IncSubtensor)
+def pytorch_funcify_IncSubtensor(op, node, **kwargs):
+    idx_list = getattr(op, "idx_list", None)
+
+    if getattr(op, "set_instead_of_inc", False):
+
+        def torch_fn(x, indices, y):
+            check_negative_steps(indices)
+            x[indices] = y
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:66]: We also need to clone if the op is not inplace

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+            return x
+
+    else:
+
+        def torch_fn(x, indices, y):
+            check_negative_steps(indices)
+            x1 = x.clone()
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:73]: We only need to clone if the Op is not inplace

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+            x1[indices] += y
+            return x1
+
+    def incsubtensor(x, y, *ilist, torch_fn=torch_fn, idx_list=idx_list):
+        indices = indices_from_subtensor(ilist, idx_list)
+        if len(indices) == 1:
+            indices = indices[0]
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:80]: Why is this needed? Can't we just call  `x[tuple(indices)]` regardless of whether its one or more?

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+
+        return torch_fn(x, indices, y)
+
+    return incsubtensor
+
+
+@pytorch_funcify.register(AdvancedIncSubtensor)
+@pytorch_funcify.register(AdvancedIncSubtensor1)
+def pytorch_funcify_AdvancedIncSubtensor(op, node, **kwargs):
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:89]: Same comments as above for IncSubtensor

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+
+    else:
+
+        def torch_fn(x, indices, y):
+            check_negative_steps(indices)
+            x1 = x.clone()
+            x1[indices] += y
+            return x1
+
+    def incsubtensor(x, y, *ilist, torch_fn=torch_fn, idx_list=idx_list):
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:77]: Not needed?

```suggestion
    def incsubtensor(x, y, *ilist):
```

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+        indices = indices_from_subtensor(ilist, idx_list)
+        if len(indices) == 1:
+            indices = indices[0]
+
+        return torch_fn(x, indices, y)
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:82]: Can check negative steps here since it's common to both cases

File: tests/link/pytorch/test_subtensor.py [hunk 29]
```diff
@@ -0,0 +1,227 @@
...
+    with pytest.raises(
+        NotImplementedError, match="Negative step sizes are not supported in Pytorch"
+    ):
+        out_pt = x_pt[[1, 2], ::-1]
+        out_fg = FunctionGraph([x_pt], [out_pt])
+        assert isinstance(out_pt.owner.op, pt_subtensor.AdvancedSubtensor)
+        compare_pytorch_and_py(out_fg, [x_np])
+
+
+def test_pytorch_SetSubtensor():
```
Comment (ricardoV94) [inline tests/link/pytorch/test_subtensor.py:112]: Parametrize test to reuse code to test Set and Incsubtensor with same function

File: tests/link/pytorch/test_subtensor.py [hunk 30]
```diff
@@ -0,0 +1,227 @@
...
+    compare_pytorch_and_py(out_fg, [x_test])
+
+    st_pt = pt.as_tensor_variable(np.r_[-1.0, 0.0].astype(config.floatX))
+    out_pt = pt_subtensor.inc_subtensor(x_pt[:2, 0, 0], st_pt)
+    assert isinstance(out_pt.owner.op, pt_subtensor.IncSubtensor)
+    out_fg = FunctionGraph([x_pt], [out_pt])
+    compare_pytorch_and_py(out_fg, [x_test])
+
+
+def test_pytorch_AvdancedIncSubtensor():
```
Comment (ricardoV94) [inline tests/link/pytorch/test_subtensor.py:188]: Parametrize inc and set subtensor with same code inside

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 31]
```diff
@@ -0,0 +1,111 @@
...
+@pytorch_funcify.register(AdvancedSubtensor)
+def pytorch_funcify_AdvSubtensor(op, node, **kwargs):
+    def advsubtensor(x, *indices):
+        check_negative_steps(indices)
+        return x[indices]
+
+    return advsubtensor
+
+
+@pytorch_funcify.register(MakeSlice)
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:50]: Nitpick this should come before the AdvancedSubtensor dispatch as that also makes use of MakeSlice

**Review by ricardoV94 — COMMENTED** [review]
Getting there, some smaller feedback this time

File: pytensor/link/pytorch/dispatch/subtensor.py (hunk 24 above)
Comment (HarshvirSandhu) [reply pytensor/link/pytorch/dispatch/subtensor.py:73]: The `IncSubtensor` tests fails without cloning, in `compare_pytorch_and_py` the inputs the `pytensor_fn` get are already incremented by the `pytorch_fn`

Comment (ricardoV94) [reply pytensor/link/pytorch/dispatch/subtensor.py:73]: You have to check the inplace property of the Op. When true you don't need to copy, otherwise you do. This is also the case for AdvancedIncSubtensor

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 32]
```diff
@@ -0,0 +1,120 @@
...
+
+    if getattr(op, "set_instead_of_inc", False):
+
+        def torch_fn(x, indices, y):
+            if op.inplace:
+                x[tuple(indices)] = y
+                return x
+            x1 = x.clone()
+            x1[tuple(indices)] = y
+            return x1
```
[the remark this answers is not in the document]
Comment (HarshvirSandhu) [reply pytensor/link/pytorch/dispatch/subtensor.py:70]: No,  the tests fail
In `compare_pytorch_and_py` the inputs of `pytensor_fn` get are already incremented by `pytorch_fn`

**Comment by ricardoV94** [comment]
@HarshvirSandhu I pushed some simplification and also needed special logic for repeated indices in IncSubtensor. Feel free to review and see if you spot any issues. Note that I squashed the commits and force-pushed, so you'll have to retrieve the remote branch before you can do any changes

File: pytensor/link/pytorch/dispatch/subtensor.py [hunk 33]
```diff
@@ -0,0 +1,124 @@
...
+    inplace = op.inplace
+    if op.set_instead_of_inc:
+
+        def set_subtensor(x, y, *flattened_indices):
+            indices = indices_from_subtensor(flattened_indices, idx_list)
+            check_negative_steps(indices)
+            if not inplace:
+                x = x.clone()
+            x[indices] = y
+            return x
```
Comment (ricardoV94) [inline pytensor/link/pytorch/dispatch/subtensor.py:65]: This works for me locally, are you sure it wasn't working on your end?

**Review by ricardoV94 — APPROVED** [review]
